### DISCUSSION TITLE
Freeing buffer memory vs. corrupted output images

### DISCUSSION THREAD

**Discussion by kstanikviacbs** [libvips/libvips#2783] [Q&A] [answered]
Hi again! 

I'd like to clarify one thing I'm not sure about - is freeing buffer with ```g_free(buf)``` after writing to buffer safe in terms of output image corruption? There seem to be few examples here and there in the repository doing it that way (https://github.com/libvips/libvips/issues/1931#issuecomment-744698646) while there's also a suggestion in docs and few topics mentioning that safest way would be to connect with Glib to ```"postclose"``` signal (https://github.com/libvips/libvips/blob/master/libvips/foreign/jpegload.c#L532). I would appreciate clarification. 

Thanks!

**Comment by jcupitt** [comment, marked as answer]
Hi again @kstanikviacbs,

I guess this is regarding a buffer that you pass to `new_from_buffer()`?

Yes, the pointer needs to stay valid for as long as you are using (directly or indirectly) the image you make from it. The simplest thing is to attach to the `postclose` signal on the image and free there.

For `write_to_buffer()`, the pointer you get is yours (libvips is no longer using it) and you can free it any time you like.

**Reply by versi-tech** [reply]
You're right, I mixed things up a bit. Understood, thanks for clarification. Clear and concise explanation as always.
